yarn-project/types/src/contract_data.ts

```diff
@@ -16,14 +68,25 @@ export class ContractData {
      * The L1 address of the contract, (20 bytes).
      */
     public portalContractAddress: EthAddress,
-  ) {}
+
+    /**
+     * ABIs of public functions
+     */
+    public publicFunctions?: EncodedContractFunction[],
+  ) {
+    if (publicFunctions?.length) {
+      this.bytecode = serializeBufferArrayToVector(publicFunctions.map(fn => fn.toBuffer()));
+    }
+  }
 
   /**
    * Serializes this instance into a buffer, using 20 bytes for the eth address.
    * @returns Encoded buffer.
    */
   public toBuffer(): Buffer {
-    return serializeToBuffer(this.contractAddress, this.portalContractAddress.toBuffer());
+    return this.bytecode
+      ? serializeToBuffer(this.contractAddress, this.portalContractAddress.toBuffer(), this.bytecode)
+      : serializeToBuffer(this.contractAddress, this.portalContractAddress.toBuffer());
```

Comment:
Shouldn't we append a zero at the end of this branch, representing the size of the bytecode? Otherwise the toBuffer and fromBuffer methods are not symmetric, since fromBuffer will always try to ready the length of the vector after the portal contract address.